rvgo/fast/instrumented.go

```diff
@@ -73,7 +73,7 @@ func (m *InstrumentedState) Step(proof bool) (wit *StepWitness, err error) {
 
 func (m *InstrumentedState) readPreimage(key [32]byte, offset uint64) (dat [32]byte, datLen uint64, err error) {
 	preimage := m.lastPreimage
-	if key != m.lastPreimageKey {
+	if preimage == nil || key != m.lastPreimageKey {
```

Comment:
I'm not sure this is right - cannon doesn't perform this check - see [here](https://github.com/ethereum-optimism/optimism/blob/9d41b8cb98769a9694438ed835779e9f205accab/cannon/mipsevm/exec/preimage.go#L48-L65).

Reply:
Yep, this is normally not needed, but the reviewers pointed out that this can result in NPE if a preimage read is not preceded with preimage write - just in case the client program misbehaves.

Reply:
Also added a test here: https://github.com/ethereum-optimism/asterisc/pull/114/commits/1203fc428159d865b88012e372cce2377fe7e3c1 so that we can check the regression